File: src/main/java/com/tms/csp/transforms/ImpsToCnf.java
package com.tms.csp.transforms;

import com.tms.csp.ast.Exp;

/**
 * Replaces each imp(a b) with an or(!a b)
 */
public class ImpsToCnf extends BaseTransformer {

    @Override
    protected Exp local(Exp in) {
        return in.asImp().toOr();
    }

    @Override
    protected boolean executeLocal(Exp in) {
        return in.isImp();
    }


}

File: src/main/java/com/tms/csp/transforms/ImpsToCnf.kt
package com.tms.csp.transforms

import com.tms.csp.ast.Exp

/**
 * Replaces each imp(a b) with an or(!a b)
 */
class ImpsToCnf : BaseTransformer() {

    override fun local(inExp: Exp): Exp? {
        return inExp.asImp().toOr
    }

    override fun executeLocal(inExp: Exp): Boolean {
        return inExp.isImp
    }


}
